File: UCFPT/highschoolonline2021/solutions/chocolate/chocolate.py

# Read in number of contests
contests = int(input())

# Loop over each contest
for i in range (0, contests):
   # Read in highest judge ID and number of problems
   n, m = map(int, input().split())

   # Track starts and ends
   starts = [0] * (n + 2)
   ends = [0] * m

   # Loop over the problems
   for j in range (0, m):
      # Read the range
      left, right = map(int, input().split())

      # Track how many intervals start at the "left" value
      starts[left] += 1;

      # Then, just store for each problem where its interval ends
      ends[j] = right;

   # Start by assuming all ranges cannot be combined
   result = m

   # Then, check each interval end and see if it "runs up" against
   # any starting interval (if so, we can combine this interval)
   for j in range (0, m):
      # See if an interval starts right after this one ends
      if starts[ends[j]+1] > 0:
         # If so, decrement the range count and "combine" the start
         # by reducing it by one in count
         result -= 1
         starts[ends[j]+1] -= 1

   # Print the answer!
   print(result)

File: UCFPT/highschoolonline2021/solutions/chocolate/chocolate.cpp
//Chocolate.cpp
//Jacob Steinebronn

#include <bits/stdc++.h>
#define newl '\n'

using namespace std;

int main(){
    int nq;
    cin >> nq;

    while(nq--){
        int n, m;
        cin >> n >> m;

        //Keep track for each possible starting point, how many ranges start at this point
        vector<int> numStartHere(n+2,0);

        //Keep track of each ending point
        vector<int> ends(m,0);

        //Parse input
        for(int i=0;i<m;i++){
            int l, r;
            cin >> l >> r;
            numStartHere[l]++;
            ends[i]=r;
        }

        //Assume none of the ranges can be combined together
        int ans = m;
        for(int end:ends){
            //If a range starts right after this current one ends, we can merge them, so decrement ans since 2 ranges became one
            if(numStartHere[end+1]){
                numStartHere[end+1]--;
                ans--;
            }
        }
        cout << ans << newl;

    }
}
